compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeNpuDmaCpyNd.cpp

```diff
@@ -0,0 +1,187 @@
+          dmaOp.emitOpError()
+              << "has source in L3, but does not have source addressing. "
+                 "Source addressing is required to canonicalize to rank-4 "
+                 "here.";
+          return WalkResult::interrupt();
+        }
+        // Make rank-4.
+        srcOffsets = getPrepended(srcOffsets, zero);
+        srcSizes = getPrepended(srcSizes, one);
+        srcStrides = getPrepended(srcStrides, zero);
+
+        // Ensure that the zero stride is in the outer-most dimension.
+        std::optional<uint32_t> maybeSwapIndex =
+            getZeroStrideIndex(srcSizes, srcStrides, dmaOp);
+        if (!maybeSwapIndex.has_value()) {
+          return WalkResult::interrupt();
+        }
+        uint32_t swapIndex = maybeSwapIndex.value();
+        std::swap(srcOffsets[0], srcOffsets[swapIndex]);
+        std::swap(srcSizes[0], srcSizes[swapIndex]);
+        std::swap(srcStrides[0], srcStrides[swapIndex]);
+      }
+
+      if (dmaOp.getTargetMemorySpaceAsUInt() == 0) {
+        if (!dmaOp.hasTargetAddressing()) {
+          dmaOp.emitOpError()
+              << "has target in L3, but does not have target addressing. "
+                 "Target addressing is required to canonicalize to rank-4 "
+                 "here.";
+          return WalkResult::interrupt();
+        }
+
+        // Make rank-4.
+        tgtOffsets = getPrepended(tgtOffsets, zero);
+        tgtSizes = getPrepended(tgtSizes, one);
+        tgtStrides = getPrepended(tgtStrides, zero);
+
+        // Ensure that the zero stride is in the outer-most dimension.
+        std::optional<uint32_t> maybeSwapIndex =
+            getZeroStrideIndex(tgtSizes, tgtStrides, dmaOp);
+        if (!maybeSwapIndex.has_value()) {
+          return WalkResult::interrupt();
+        }
+        uint32_t swapIndex = maybeSwapIndex.value();
+        std::swap(tgtOffsets[0], tgtOffsets[swapIndex]);
+        std::swap(tgtSizes[0], tgtSizes[swapIndex]);
+        std::swap(tgtStrides[0], tgtStrides[swapIndex]);
+      }
+
+      rewriter.setInsertionPoint(dmaOp);
+
+      // Replace the npu.dma_cpy_nd with the canonicalized version.
+      dmaOp = rewriter.replaceOpWithNewOp<AMDAIE::NpuDmaCpyNdOp>(
+          dmaOp, dmaOp.getDma(), tgtOffsets, tgtSizes, tgtStrides, srcOffsets,
+          srcSizes, srcStrides, dmaOp.getTargetBdId(), dmaOp.getSourceBdId());
+
+      return WalkResult::advance();
+    });
+
+    if (walkResult.wasInterrupted()) {
+      return signalPassFailure();
+    }
+  }
+
+ private:
+  // Repeat prepend 'def' to 'tail' to make 'tail' have 4 elements.
+  static SmallVector<OpFoldResult, 4> getPrepended(ArrayRef<OpFoldResult> tail,
+                                                   Attribute def) {
+    assert(tail.size() <= 4);
+    SmallVector<OpFoldResult, 4> res(4 - tail.size(), def);
+    for (int i = 0; i < tail.size(); ++i) {
+      res.push_back(tail[i]);
+    }
+    return res;
+  }
+
+  static std::optional<uint32_t> getZeroStrideIndex(
```

Comment:
I don't think it's clear from the name of the function that the majority of its logic is for checking that there is no non-unit dimension before the first `stride == 0` dimension. It would be good to clarify with a docstring and/or adjust the name.